### PR TITLE
[Bugfix][Spec Decode][V0] fix: update logits processor for MQA scoring

### LINKED ISSUE #9423: [Bug]: Speculative decoding breaks guided decoding.
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jul 29 2024, 16:56:48) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-1063-azure-x86_64-with-glibc2.35
Is CUDA available: N/A
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: 
GPU 0: NVIDIA A100 80GB PCIe
GPU 1: NVIDIA A100 80GB PCIe

Nvidia driver version: 535.161.08
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             48
On-line CPU(s) list:                0-47
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 7V13 64-Core Processor
CPU family:                         25
Model:                              1
Thread(s) per core:                 1
Core(s) per socket:                 48
Socket(s):                          1
Stepping:                           1
BogoMIPS:                           4890.88
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext perfctr_core invpcid_single vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves clzero xsaveerptr rdpru arat umip vaes vpclmulqdq rdpid fsrm
Hypervisor vendor:                  Microsoft
Virtualization type:                full
L1d cache:                          1.5 MiB (48 instances)
L1i cache:                          1.5 MiB (48 instances)
L2 cache:                           24 MiB (48 instances)
L3 cache:                           192 MiB (6 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-23
NUMA node1 CPU(s):                  24-47
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Mitigation; safe RET, no microcode
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.21.5
[conda] Could not collect
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: N/A
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
[4mGPU0	GPU1	NIC0	CPU Affinity	NUMA Affinity	GPU NUMA ID[0m
GPU0	 X 	NV12	NODE	0-23	0		N/A
GPU1	NV12	 X 	SYS	24-47	1		N/A
NIC0	NODE	SYS	 X 				

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_an0
```

</details>

### Model Input Dumps

_No response_

### 🐛 Describe the bug

I run the vLLM server with speculative decoding as follows:
```.py
NCCL_GRAPH_FILE="/home/azureuser/apps/ai-kyc_vu/vllm_server/gpu_graph.xml" python -m vllm.entrypoints.openai.api_server
--model "neuralmagic/Meta-Llama-3.1-70B-Instruct-FP8"
--served-model-name ""
--guided-decoding-backend "outlines"
--gpu-memory-utilization 0.9 
--port 7999
--worker_use_ray
--max-model-len 40000
--tensor-parallel-size 2
--speculative-model="[ngram]"
--num_speculative_tokens 5
--ngram_prompt_lookup_max=4
--use-v2-block-manager
```

I then prompt the LLM with guided json created from the following pydantic model:

```.py
class A(BaseModel):
    a: int
```

I  incorporate the guided json into the  following testing prompt:
```.json
{
    "model": "",
    "messages": [
      {
        "role": "system",
        "content": ""
      },
      {
        "role": "user",
        "content": "How are you?"
      }
    ],
    "guided_decoding_backend": "outlines",
    "guided_json": {"properties": {"a": {"title": "A", "type": "integer"}}, "required": ["a"], "title": "A", "type": "object"},
    "max_tokens": 200,
    "top_k": 1,
    "stream": false
}
```

The json guidance should ensure that the output is a valid json object. However, vLLM returns the following incomplete json object:
```
{"a": 1
```

When I disable ngram-speculative decoding, the same prompt works, and returns a complete json object:
```
{"a": 1}
```
    
This means that somehow, ngram-speculative decoding breaks json guidance.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by llsj14**
FIX #9423 

## Summary
- In issue #9423, the MQA scorer sampled different tokens than the BatchExpansion scorer, even when using the greedy option.
- The model itself produced the same hidden states, but differences in `model_input.sampling_metadata` led to discrepancies.
- The sampling metadata generated by the MQA scorer is not compatible with the operation of [LogitsProcessor](https://github.com/vllm-project/vllm/blob/5f671cb4c3145194e94ffb393ee459432f7fa2b8/vllm/model_executor/layers/logits_processor.py#L78-L80).
- Therefore, this PR aims to make the LogitsProcessor work correctly with the MQA scorer. I updated the LogitsProcessor because handling the sampling metadata from the MQA scorer is not straightforward.

https://github.com/vllm-project/vllm/blob/2079e43beecc486a607c9d79ab691e0e4563aa11/vllm/worker/model_runner.py#L1762-L1763

<model_input.sampling_metadata from MQA Scorer>
```
SamplingMetadata(seq_groups=[

SequenceGroupToSample(seq_ids=[1], sampling_params=SamplingParams(n=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.0, temperature=1.0, top_p=1.0, top_k=1, min_p=0.0, seed=None, stop=[], stop_token_ids=[], bad_words=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=32, min_tokens=0, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None, guided_decoding=None), seq_data={1: SequenceData(prompt_token_ids=array('l', [1, 3087, 8970, 338, 263]), output_token_ids=(29912, 29871, 29908, 29874, 1115, 29871, 29908, 29874, 1115), cumulative_logprob=0.0, get_num_computed_tokens=10},seq_len=None, query_len=4, generator=None, is_prompt=False, prompt_logprob_indices=[], sample_indices=[0, 1, 2, 3])],

selected_token_indices=tensor([0, 1, 2, 3], device='cuda:0'), categorized_sample_indices={<SamplingType.GREEDY: 0>: tensor([], device='cuda:0', dtype=torch.int32), <SamplingType.RANDOM: 1>: tensor([0, 1, 2, 3], device='cuda:0', dtype=torch.int32), <SamplingType.RANDOM_SEED: 2>: tensor([], device='cuda:0', dtype=torch.int32)})
```

<model_input.sampling_metadata from BatchExpansion Scorer>
```
SamplingMetadata(seq_groups=[

SequenceGroupToSample(seq_ids=[1], sampling_params=SamplingParams(n=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.0, temperature=1.0, top_p=1.0, top_k=1, min_p=0.0, seed=None, stop=[], stop_token_ids=[], bad_words=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=32, min_tokens=0, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None, guided_decoding=None), seq_data={1: SequenceData(prompt_token_ids=array('l', [1, 3087, 8970, 338, 263]), output_token_ids=(29912, 29871, 29908, 29874, 1115, 29871), cumulative_logprob=0.0, get_num_computed_tokens=10}, seq_len=None, query_len=1, generator=None, is_prompt=False, prompt_logprob_indices=[], sample_indices=[0]), 
SequenceGroupToSample(seq_ids=[2], sampling_params=SamplingParams(n=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.0, temperature=1.0, top_p=1.0, top_k=1, min_p=0.0, seed=None, stop=[], stop_token_ids=[], bad_words=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=32, min_tokens=0, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None, guided_decoding=None), seq_data={2: SequenceData(prompt_token_ids=array('l', [1, 3087, 8970, 338, 263]), output_token_ids=(29912, 29871, 29908, 29874, 1115, 29871, 29908), cumulative_logprob=0.0, get_num_computed_tokens=11}, seq_len=None, query_len=1, generator=None, is_prompt=False, prompt_logprob_indices=[], sample_indices=[1]), 
SequenceGroupToSample(seq_ids=[3], sampling_params=SamplingParams(n=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.0, temperature=1.0, top_p=1.0, top_k=1, min_p=0.0, seed=None, stop=[], stop_token_ids=[], bad_words=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=32, min_tokens=0, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None, guided_decoding=None), seq_data={3: SequenceData(prompt_token_ids=array('l', [1, 3087, 8970, 338, 263]), output_token_ids=(29912, 29871, 29908, 29874, 1115, 29871, 29908, 29874), cumulative_logprob=0.0, get_num_computed_tokens=12}, seq_len=None, query_len=1, generator=None, is_prompt=False, prompt_logprob_indices=[], sample_indices=[2]),
SequenceGroupToSample(seq_ids=[4], sampling_params=SamplingParams(n=1, presence_penalty=0.0, frequency_penalty=0.0, repetition_penalty=1.0, temperature=1.0, top_p=1.0, top_k=1, min_p=0.0, seed=None, stop=[], stop_token_ids=[], bad_words=[], include_stop_str_in_output=False, ignore_eos=False, max_tokens=32, min_tokens=0, logprobs=None, prompt_logprobs=None, skip_special_tokens=True, spaces_between_special_tokens=True, truncate_prompt_tokens=None, guided_decoding=None), seq_data={4: SequenceData(prompt_token_ids=array('l', [1, 3087, 8970, 338, 263]), output_token_ids=(29912, 29871, 29908, 29874, 1115, 29871, 29908, 29874, 1115), cumulative_logprob=0.0, get_num_computed_tokens=13}, seq_len=None, query_len=1, generator=None, is_prompt=False, prompt_logprob_indices=[], sample_indices=[3])], 

selected_token_indices=tensor([0, 1, 2, 3], device='cuda:0'), categorized_sample_indices={<SamplingType.GREEDY: 0>: tensor([], device='cuda:0', dtype=torch.int32), <SamplingType.RANDOM: 1>: tensor([0, 1, 2, 3], device='cuda:0', dtype=torch.int32), <SamplingType.RANDOM_SEED: 2>: tensor([], device='cuda:0', dtype=torch.int32)}), 
```

### Run Server
```
python -m vllm.entrypoints.openai.api_server \
     --model meta-llama/Llama-2-7b-hf \
     --port 8087 \
     --disable-custom-all-reduce \
     --swap-space 0 \
     --gpu-memory-utilization 0.8 \
     --speculative_model "[ngram]" \
     --ngram_prompt_lookup_max 2 \
     --num_speculative_tokens 3 \
     --guided-decoding-backend "outlines" \
     --enforce-eager \
```

### Request
```
curl http://localhost:8087/v1/completions \
    -H "Content-Type: application/json" \
    -d '{
        "model": "meta-llama/Llama-2-7b-hf",
        "prompt": "San Francisco is a",
                "max_tokens": 32,
                "top_k": 1,
                "top_p": 1.0,
        "temperature": 1.0,
                "guided_decoding_backend": "outlines",
                "guided_json": {"properties": {"a": {"title": "A", "type": "integer"}}, "required": ["a"], "title": "A", "type": "object"}
    }' | jq
```

### Logits
- as-is
```
tensor([[    -inf,     -inf,   3.6250,  ...,     -inf,     -inf,     -inf],
        [-10.6406, -11.0078,   1.7412,  ...,  -0.7197,  -6.2422,  -2.6250],
        [-11.1641,  -9.0703,   1.0527,  ...,  -4.8320,  -8.4844,  -7.2109],
        [ -9.1719,  -6.3828,   4.4492,  ...,  -3.3496,  -3.8672,  -2.7852]],
       device='cuda:0', dtype=torch.float16)
```

- to-be
```
tensor([[  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
        [  -inf,   -inf, 1.7412,  ...,   -inf,   -inf,   -inf],
        [  -inf,   -inf, 1.0527,  ...,   -inf,   -inf,   -inf],
        [  -inf,   -inf, 4.4492,  ...,   -inf,   -inf,   -inf]],
       device='cuda:0', dtype=torch.float16)
```

### Response
- as-is
```
{"a": 1
```
- to-be
```
{ \"a\": 1 }
```

### CODE REVIEW

**Comment by llsj14** [comment]
@roberthoenig @LiuXiaoxuanPKU 
This PR aims to fix the bug when computing multiple logits with sampling metadata from MQA Scorer.
Please take a look when you get a chance.

File: vllm/model_executor/layers/logits_processor.py [hunk 1]
```diff
@@ -141,10 +142,17 @@ def _apply_logits_processors(
...
-                                              seq_group.sample_indices):
+            for i, (seq_id, logits_row_idx) in enumerate(
+                    zip_longest(seq_ids,
+                                seq_group.sample_indices,
+                                fillvalue=seq_ids[-1])):
                 logits_row = logits[logits_row_idx]
-                past_tokens_ids = seq_group.seq_data[seq_id].output_token_ids
+                max_output_len = len(
+                    seq_group.seq_data[seq_id].output_token_ids)
+                num_samples = len(seq_group.sample_indices)
```
Comment (benchislett) [inline vllm/model_executor/layers/logits_processor.py:152]: QQ: Are sample_indices guaranteed to be contiguous (0, 1, 2, .., N) ?

Comment (llsj14) [reply vllm/model_executor/layers/logits_processor.py:152]: Yes, sample_indices is always contiguous because it is created using the range.

https://github.com/vllm-project/vllm/blob/f17f1d46086692a2973fad94860a95799fbd8582/vllm/model_executor/sampling_metadata.py#L335

Comment (llsj14) [reply vllm/model_executor/layers/logits_processor.py:152]: I added a comment shortly to maintain that requirement in the future.

**Comment by llsj14** [comment]
@sroy745 @LiuXiaoxuanPKU 
Would you also check this PR when you have time please?

**Comment by llsj14** [comment]
@LiuXiaoxuanPKU 
Do you also plan to support the MQA scorer with v1?
If so, I assume this bug might affect v1 as well.
